### PR TITLE
src: make Suricata build on Solaris

### PR DESCRIPTION
**Pull request by l1gi**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [X] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)
- [X] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues
      (if applicable)

Link to ticket: https://redmine.openinfosecfoundation.org/issues/7005

Describe changes:
- src/main.c: Fix alignment problems on SPARC. Based on similar fixes found at:
  https://lists.debian.org/debian-sparc/2012/01/msg00012.html
and:
  https://www.winehq.org/pipermail/wine-patches/2011-February/098547.html
- src/source-pcap-file-directory-helper.c, src/util-path.c: Solaris doesn't define d_type dirent member and DT_REG type.
- src/suricata.c: Only Linux and the BSDs support the RLIMIT_NPROC resource limit.
- src/util-syslog.c: Solaris syslog.h doesn't define LOG_FTP.
- src/util-time.c: Solaris tm doesn't define tm_gmtoff member.

### CODE REVIEW

**Comment by l1gi** [comment]
> What's the difference between `__sun` and `__sparc`?

Hi Jeff,

`__sun` means SunOS, in other words Solaris OS.
`__sparc` means hardware platform with SPARC CPU inside.

That part relates just to SPARC. I have hit a memory miss-align. That is quite common and that `ta 6` is the easiest solution/workaround. Once the CPU detects the address alignment issue, that instruction just causes CPU to emulate that particular instruction with multiple instructions. There is slight hardware penalty, but it works reliably.

Regards,
m.

**Comment by l1gi** [comment]
```
Run ./scripts/clang-format.sh check-branch --diffstat --show-commits  >> check_formatting_log.txt 2>&1
 src/threads.h     |  4 +---
 src/util-path.c   |  2 +-
 src/util-syslog.c | 32 ++++++++------------------------
 3 files changed, 10 insertions(+), 28 deletions(-)

Commits on branch (new -> old):
23bbfe2d3 src: make Suricata build on Solaris

ERROR: Branch requires formatting
ERROR: View required changes with: clang-format.sh check-branch --diff
ERROR: Use ./scripts/clang-format.sh branch to fix formatting,
            then add formatting changes to a new commit
Error: Formatting is not following code style guide!
Error: Process completed with exit code 1.
```

Do I understand correctly, that the commit message is wrong(ly formatted)? Could anyone guide me how to correct it?

Thank you,
m.

**Comment by jlucovsky** [comment]
> ```
> Run ./scripts/clang-format.sh check-branch --diffstat --show-commits  >> check_formatting_log.txt 2>&1
>  src/threads.h     |  4 +---
>  src/util-path.c   |  2 +-
>  src/util-syslog.c | 32 ++++++++------------------------
>  3 files changed, 10 insertions(+), 28 deletions(-)
> 
> Commits on branch (new -> old):
> 23bbfe2d3 src: make Suricata build on Solaris
> 
> ERROR: Branch requires formatting
> ERROR: View required changes with: clang-format.sh check-branch --diff
> ERROR: Use ./scripts/clang-format.sh branch to fix formatting,
>             then add formatting changes to a new commit
> Error: Formatting is not following code style guide!
> Error: Process completed with exit code 1.
> ```
> 
> Do I understand correctly, that the commit message is wrong(ly formatted)? Could anyone guide me how to correct it?
> 
> Thank you, m.

This should help: https://docs.suricata.io/en/latest/devguide/codebase/code-style.html

**Comment by jlucovsky** [comment]
> > What's the difference between `__sun` and `__sparc`?
> 
> Hi Jeff,
> 
> `__sun` means SunOS, in other words Solaris OS. `__sparc` means hardware platform with SPARC CPU inside.
> 
> That part relates just to SPARC. I have hit a memory miss-align. That is quite common and that `ta 6` is the easiest solution/workaround. Once the CPU detects the address alignment issue, that instruction just causes CPU to emulate that particular instruction with multiple instructions. There is slight hardware penalty, but it works reliably.
> 
> Regards, m.

Is there a use case when `__sun` is defined but `__sparc` isn't?

File: src/main.c [hunk 1]
```diff
@@ -19,6 +19,14 @@
 
 int main(int argc, char **argv)
 {
+#ifdef __sparc
```
Comment (victorjulien) [inline src/main.c:22]: I would like to see this else where, like in util-cpu.c/h and then be a single call from main

**Review by victorjulien — CHANGES_REQUESTED** [review]
See inline comment. Other changes make sense to me.

**Comment by l1gi** [comment]
> Is there a use case when `__sun` is defined but `__sparc` isn't?

Yes, Solaris supports all kind of Intel architecture like `__i386`, `__amd64` but also `__x86_64`: [GCC architectures](https://sourceforge.net/p/predef/wiki/Architectures/).

**Comment by l1gi** [comment]
Sorry for an extra commit.

File: src/main.c [hunk 2]
```diff
@@ -19,6 +19,10 @@
...
+    EnableSparcMisalignEmulation();
```
Comment (victorjulien) [inline src/main.c:23]: Would like to see this be more generic, like a single call out to something like `UtilCpuSpecificInit()` where inside that (in util-cpu.c) you would have this ifdef'd call to `EnableSparcMisalignEmulation`

**Comment by catenacyber** [comment]
friendly ping @l1gi will you work further on this ?

**Comment by l1gi** [comment]
> friendly ping @l1gi will you work further on this ?

Yes, sorry for delay. Will try to prepare the new change today or tomorrow.

Thank you,
m.

**Comment by catenacyber** [comment]
Thanks for the changes, we expect a new PR instead of pushing on the existing one cf https://docs.suricata.io/en/latest/devguide/contributing/code-submission-process.html#pull-requests

**Comment by catenacyber** [comment]
CI is red, please open a new rebased PR

**Pull request closed without merging**